R/PipeOpDateFeatures.R

```diff
@@ -0,0 +1,209 @@
+#' features such as seconds, i.e., second 21 and second 22 are one second apart, but so are second
+#' 60 and second 01. The transformation always assumes that `min(x) = 0`, therefore prior shifting
+#' the values internally by minus one may occurr if necessary.
+#'
+#' @section Methods:
+#' Only methods inherited from [`PipeOpTaskPreprocSimple`]/[`PipeOp`].
+#'
+#' @examples
+#' library("mlr3")
+#' dat = iris
+#' set.seed(1)
+#' dat$date = sample(seq(from = as.POSIXct("2020-01-29"), to = as.POSIXct("2020-02-28"), by = "hour"),
+#'   size = 150L)
+#' task = TaskClassif$new("iris_date", backend = dat, target = "Species")
+#' pop = po("datefeatures", param_vals = list(date_variable = "date", cyclic = FALSE, minute = FALSE,
+#'   second = FALSE))
+#' pop$train(list(task))
+#' pop$state
+#' @family PipeOps
+#' @include PipeOpTaskPreproc.R
+#' @export
+PipeOpDateFeatures = R6Class("PipeOpDateFeatures",
+  inherit = PipeOpTaskPreprocSimple,
+  public = list(
+    initialize = function(id = "datefeatures", param_vals = list()) {
+      ps = ParamSet$new(params = list(
+        ParamUty$new("date_variable", custom_check = check_character, tags = "required"),
+        ParamLgl$new("cyclic", default = FALSE),
+        ParamLgl$new("year", default = TRUE, tags = "date_feature"),
+        ParamLgl$new("month", default = TRUE, tags = "date_feature"),
+        ParamLgl$new("week_of_year", default = TRUE, tags = "date_feature"),
+        ParamLgl$new("day_of_year", default = TRUE, tags = "date_feature"),
+        ParamLgl$new("day_of_month", default = TRUE, tags = "date_feature"),
+        ParamLgl$new("day_of_week", default = TRUE, tags = "date_feature"),
+        ParamLgl$new("hour", default = TRUE, tags = "date_feature"),
+        ParamLgl$new("minute", default = TRUE, tags = "date_feature"),
+        ParamLgl$new("second", default = TRUE, tags = "date_feature"),
+        ParamLgl$new("is_day", default = TRUE, tags = "date_feature")
+      ))
+      ps$values$date_variable = "date" # needed for mlr_pipeops overview
+      super$initialize(id = id, param_set = ps, param_vals = param_vals)
+    },
+
+    get_state = function(task) {
+      date_variable = self$param_set$values$date_variable
+      features = self$param_set$ids(tags = "date_feature") # default = TRUE for all date features
+      dropped = names(which(unlist(self$param_set$get_values(tags = "date_feature")) == FALSE))
+      features = features[features %nin% dropped]
+      cyclic = ifelse(is.null(self$param_set$values$cyclic), yes = self$param_set$default$cyclic,
+        no = self$param_set$values$cyclic)
+      list(date_variable = date_variable, cyclic = cyclic, features = features)
+    },
+
+    transform = function(task) {
+      dates = task$data()[[self$state$date_variable]]
+      assert_posixct(dates)
+      features = self$state$features
+      date_features = as.data.table(lapply(features, function(feature) {
```

Comment:
Can you use more idiomatic `data.table` syntax to make things mroe legible?
https://stackoverflow.com/questions/11680579/assign-multiple-columns-using-in-data-table-by-group

Reply:
fixed (?)